File: Cpp/209.cpp
#include<iostream>
#include<vector>
#include<algorithm>

using namespace std;

class Solution{
public:
	int minSubArrayLen(int s,vector<int>& nums){
		if(nums.size()==0)
			return 0;
		int i=0,j=0;
		int minLen=INT_MAX;
		int sum=0;
		while(j<nums.size()){
			sum+=nums[j++];
			while(sum>=s){
				minLen=min(minLen,j-i);
				sum-=nums[i++];
			}
		}	
		return minLen==INT_MAX?0:minLen;
	}
};

//easy to understand
class Solution2 {
public:
    int minSubArrayLen(int s, vector<int>& nums) {
        int sum=0;
        int minLen=INT_MAX;
        for(int i=0,j=-1;i<nums.size();++i){
            sum+=nums[i];
            while(sum>=s){
                minLen=min(minLen,i-j);
                j++;
                sum-=nums[j];
            }
        }
        return minLen==INT_MAX?0:minLen;
    }
};

int main(){
	vector<int> arr={2,3,1,2,4,3};
	int len=7;
	Solution s;
	cout<<s.minSubArrayLen(len,arr)<<endl;

	return 0;
}

File: Python/209.py
class Solution(object):
    def minSubArrayLen(self, s, nums):
        """
        :type s: int
        :type nums: List[int]
        :rtype: int
        """
        sum, n = 0, len(nums)
        i,j, minLen = -1, 0, n+1
        while j<n:
            sum += nums[j]
            while sum >= s:
                minLen = min(minLen, j-i)
                i+=1
                sum -=nums[i]
            j+=1
        return minLen if minLen != n+1 else 0
                
